File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {

    public static double[] solution(int k, int[][] ranges) {

        List<Double> areas = new ArrayList<>();
        int init = k;

        while (k > 1) {
            if (k % 2 == 0) {
                k /= 2;
            } else {
                k = k * 3 + 1;
            }

            double area = (double) (k + init) / 2;
            areas.add(area);
            init = k;
        }

        int n = areas.size();
        double[] answer = new double[ranges.length];

        for (int i = 0; i < ranges.length; i++) {
            int[] range = ranges[i];

            int start = range[0];
            int end = 0;
            if(range[1] <= 0) {
                end = n + range[1];
            }

            if(start < end) {
                for (int j = start; j < end; j++) {
                    answer[i] += areas.get(j);
                }
            }else if(start > end){
                answer[i] = -1;
            }
        }
        return answer;
    }

}

File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.cpp
#include <string>
#include <vector>

using namespace std;

vector<double> solution(int k, vector<vector<int>> ranges) {
    vector<int> Collatz;
    int count = 0;
    while(true)
    {
        Collatz.push_back(k);
        
        if (k == 1)
            break;
        
        if (k % 2 == 0)
        {
            k /= 2;
        }
        else
        {
            k = k * 3 + 1;
        }
        
        ++count;
    }
    
    vector<double> areas;
    for(int i = 0; i < Collatz.size() - 1; ++i)
    {
        int upperWidth = Collatz[i];
        int lowerWidth = Collatz[i + 1];

        areas.push_back((upperWidth + lowerWidth) / 2.f);
    }
    
    vector<double> answer;
    
    for(int i = 0; i < ranges.size(); ++i)
    {
        int start = ranges[i][0];
        int end = count + ranges[i][1];
        
        double area = 0.f;
        
        if (end < start)
            area = -1.f;
        else if (start < end)
        {
            for(int j = start; j < end; ++j)
            {
                area += areas[j];
            }
        }
        
        answer.push_back(area);
    }
    
    return answer;
}